### PR TITLE
feat: Generic SMT keys with `From<Bytes32>`

### PR DESCRIPTION
**Pull request by bvrooman**
Right now, SMT keys are required to be instances of `[u8; 32]`. However, this conflicts with an internal constraint on `ChildError` where the child key must be displayed in the event of an error. Because we cannot implement `Display` on `[u8; 32]`, we get a compilation error in the `fuel-core` when handling child errors. To address this, we must allow client code to use a wrapper struct that implements `Display`, as well as compatibility traits `From<[u8; 32]>` as well as `AsRef<[u8; 32]>`.

This PR replaces usages of `[u8; 32]` in the SMT public interface with a generic `Key`. `Key` must implement these compatibility traits. This will allow `fuel-core` to use its native `Bytes32` type, where these traits are already implemented.

### CODE REVIEW

**Review by freesig — CHANGES_REQUESTED** [review]
This creates a few places where references to keys / values are cloned where they were not before. It seems like the `ChildError` could simply change from `{0}` to `{0:?}` to avoid requiring the display.

**Comment by bvrooman** [comment]
It's true we can simply use `{0:?}` instead to defer to `Debug`. The implication is then that `Mappable::Key` implements `Debug`; since keys (byte arrays, u64) implement this already, we would achieve that for free. However, I chose to explore another way because I would argue that:
1. We can write superior code by allowing client libraries to use their key structs
2. Using the  `[u8; 32] Debug` loses semantics 
 
The client library should dictate the way that keys are represented at the client code level. This allows us to have more communicative Mappables. 

Instead of 
```rust
impl Mappable for ContractsAssetsMerkleData {
    type Key = [u8; 32];
    ...
}
```
We are now able to do
```rust
impl Mappable for ContractsAssetsMerkleData {
    type Key = ContractId;
    ...
}
```
And `ContractId` can specify any of its display (or debug traits). We can hex encode the contract Id, rather than printing an array of bytes, or we can write more complex displays with any sort of semantic meaning at the client level.

File: fuel-merkle/src/sparse/node.rs [hunk 1]
```diff
@@ -6,88 +6,100 @@ use crate::{
...
 #[derive(Clone)]
-pub(crate) struct Node {
+pub(crate) struct Node<Key> {
     height: u32,
     prefix: Prefix,
-    bytes_lo: Bytes32,
-    bytes_hi: Bytes32,
+    bytes_lo: Key,
+    bytes_hi: Key,
+    key: PhantomData<Key>,
```
Comment (xgreenx) [inline fuel-merkle/src/sparse/node.rs:24]: You don't need `PhantomData` because you already have `Key` field

**Review by xgreenx — COMMENTED** [review]
I like the idea of allowing any key that can be converted into `Bytes32`(maybe it is better to create a new `MerkleRoot` type).

But it seems the current implementation is overkill because instead of affecting only `impl` sections of the Merkle tree, it also affects structs like `Node` etc.

I think we should affect only the API of the Merkle Tree by this change but not the internal implementation because it should be the same as before.

**Comment by bvrooman** [comment]
> I like the idea of allowing any key that can be converted into `Bytes32`(maybe it is better to create a new `MerkleRoot` type).
> 
> But it seems the current implementation is overkill because instead of affecting only `impl` sections of the Merkle tree, it also affects structs like `Node` etc.
> 
> I think we should affect only the API of the Merkle Tree by this change but not the internal implementation because it should be the same as before.

I agree that it would be nice to keep the changes as simple as possible. I will go over this PR again to see what can be simplified. However, I do not believe we can avoid touching the internal structures. I would argue that Nodes are necessarily concerned with the Key type. I think the mental model should be something like:
- A `Leaf` consists of (Leaf Key, Value)
- A `Node` consists of (Left Key, Right Key)
- A Merkle Tree consists of many Leaves and Nodes
- A Merkle Tree is simply the public interface that manipulates Nodes using Keys

I think this model shows how changes to Key propagate through all these structures. 

For some purposes, we need to have more features available from our Keys. For things like tracing, this includes the ability to for the Key to be displayed or encoded. This is currently not possible because we use `[u8; 32]` for Keys. This represents a gap in the Merkle tree functionality. 

But as I said, I can look into simplifying the implementation. But it does not seem possible to avoid touching the Node structs themselves. Maybe I can use a concrete type that wraps `[u8; 32]` instead of a generic type, and apply conversions?

**Comment by xgreenx** [comment]
> For some purposes, we need to have more features available from our Keys. For things like tracing, this includes the ability to for the Key to be displayed or encoded

We don't need tracing for the internal details of the Merkle tree work. The Merkle tree should be a black box for `fuel-core` that works correctly and is covered by tests.

The encoding should be simple for us like `AsRef<[u8; 32]>` or `Into<[u8; 32]>`(or instead of `[u8; 32]` we can use `Bytes32` or a new type `MerkleRoot`). Because it is an internal representation of the data inside of the Merkle tree. If we need to convert internal type `[u8; 32]` into a `ContractId` for `StorageInpect` or `StorageMutate`, then it also should implement `From<[u8; 32]>`.

But it should affect only public API without affecting the nodes and leafs(internal details)

**Comment by bvrooman** [comment]
> We don't need tracing for the internal details of the Merkle tree work. The Merkle tree should be a black box for `fuel-core` that works correctly and is covered by tests.

I agree that we don't need trace the internal details of the Merkle tree. But we do need to trace errors that affect client side operations, such as a key not found in the database during `update`. This could represent a corrupted database or some other tree traversal failure. These things represent real error cases to the user, not internal details.

> The encoding should be simple for us like `AsRef<[u8; 32]>` or `Into<[u8; 32]>`(or instead of `[u8; 32]` we can use `Bytes32` or a new type `MerkleRoot`). Because it is an internal representation of the data inside of the Merkle tree. If we need to convert internal type `[u8; 32]` into a `ContractId` for `StorageInpect` or `StorageMutate`, then it also should implement `From<[u8; 32]>`.
> 
> But it should affect only public API without affecting the nodes and leafs(internal details)

It would be nice if we had could use  `AsRef<[u8; 32]>` or `Into<[u8; 32]>` and it would simplify things a lot. The only problem is that `fuel-core` keys do not implement AsRef<[u8; N]>` for their respective `N`; they implement `AsRef<[u8]>`. If we want to support these keys in `fuel-merkle`, this means that the size must be retrieved at runtime and conversions from a `&[u8]` to a `Key` must be implemented using `TryFrom`. This introduces another level of complexity with error handling.

I have tried implementing `AsRef<[u8; N]>` on these keys in `fuel-core`, but I ran into many compile errors because of the now ambiguous calls to `as_ref()`. It looks like that would be a very drastic change to implement. However, I haven't tried `Into<[u8; N]>` so maybe I can explore that

**Comment by bvrooman** [comment]
It might be possible to use `Bytes32` (or a theoretical `MerkleRoot`) from `fuel-types` directly for Keys in `fuel-merkle` instead of `[u8; 32]`. This would give us the desired features on Keys but also hopefully give us simpler conversions.

**Comment by bvrooman** [comment]
I've updated this PR to minimize changes to the `Node` and `Primitive` by changing the `MerkleTreeKey` trait to include `Into<Bytes32>`, rather than `AsRef<Bytes32>` (this  in addition to `From<Bytes32>`). This appears to work nicely, and I can do the necessary transformations from `Bytes32` to `Key`, and vice versa, fairly easily and without downstream compilation issues.
CC @xgreenx I think this should satisfy are above discussion. Let me know if there are any outstanding issues there.

**Review by xgreenx — CHANGES_REQUESTED** [review]
Okay, now I think we don't need this change =D Because we can't use `ContractId`(or other business types) as a key because the root key is a random value.

So we only need types that can represent a hash.
Why can't we use `fuel_types::Bytes32` as a default and required value?(or introduce a new `MerkleRoot` type)


The idea with generic is good, but I bring some unusual usage like: I can use `ContractId` as a key(because it implements all required traits), but the meaning is incorrect in this case. + As @freesig  mentioned, it adds a new "copy" to the code, and without proper benchmarking of it, I prefer not to add them.

**Comment by bvrooman** [comment]
> Okay, now I think we don't need this change =D Because we can't use `ContractId`(or other business types) as a key because the root key is a random value.
> 
> So we only need types that can represent a hash. Why can't we use `fuel_types::Bytes32` as a default and required value?(or introduce a new `MerkleRoot` type)
> 
> The idea with generic is good, but I bring some unusual usage like: I can use `ContractId` as a key(because it implements all required traits), but the meaning is incorrect in this case. + As @freesig mentioned, it adds a new "copy" to the code, and without proper benchmarking of it, I prefer not to add them.

An SMT is simply a function on a subset of set `X`. As long as the set `X` is closed under a hash function `f`, (in this case `sha256`) then we can calculate a root. An SMT necessarily produces values `x` inside the domain of `X`, regardless of the semantic meaning of `X`. That's exactly the point: we can calculate roots of any data set. The meaning of `ContractId` is preserved by the SMTs public interface, `update(key, value)` and `delete(key, value)`, but how it is used internally is not relevant.

That said, `Bytes32` also meets the closure property, and could work just as well. If we are uncomfortable for whatever reason with using `ContractId` here then we can use `Bytes32` instead of generics, and enforce transformations to `Bytes32` at the interface.

File: fuel-merkle/src/sparse/node.rs [hunk 2]
```diff
@@ -280,26 +281,27 @@ pub enum StorageNodeError<StorageError> {
...
-        let key = self.node.left_child_key();
-        if key == zero_sum() {
+        let key = *self.node.left_child_key();
+        if key == *zero_sum() {
             return Ok(Self::new(self.storage, Node::create_placeholder()));
         }
         let primitive = self
             .storage
-            .get(key)
+            .get(&key.into())
```
[the remark this answers is not in the document]
Comment (bvrooman) [reply fuel-merkle/src/sparse/node.rs:302]: As long as Key implements From<Bytes32> and Into<Bytes32> you can use Key here for storage. The value of the Key instance isn't really relevant.

**Comment by Voxelot** [comment]
I kind of agree with @xgreenx's line of questioning here. For example:

> A Node consists of (Left Key, Right Key)

I have a hard time picturing how these keys could relate to some domain-specific type like a `ContractId`. Even though it satisfies the trait constraints like `AsRef<[u8]>`, these merkle data keys seem like an internal implementation detail to the merkle library and likely don't need to be generic.

Taking a step back, if only the merkle library needs the ability to construct these `ChildError` responses why is it a blocker with the way things are currently by just using `hex::encode` while constructing the error?

**Comment by bvrooman** [comment]
> I kind of agree with @xgreenx's line of questioning here. For example:
> 
> > A Node consists of (Left Key, Right Key)
> 
> I have a hard time picturing how these keys could relate to some domain-specific type like a `ContractId`. Even though `ContractId` happens to satisfy trait constraints like `AsRef<[u8]>`, it wouldn't really make sense to use here. These merkle data keys seem like an internal implementation detail of the merkle library and likely don't need to be generically customizable by the consumer.

The observation is simply that an SMT can calculate a root on any subset of set `X` that is closed under `f` (i.e. sha256). As long as `f` can generate elements of `X`, we can calculate a root of a subset. In this case, we _can_ generate `ContractId`s using sha256, but, as you both have pointed out, the connection between `ContractId` and `sha256` is not really there. So it's probably best not to have this generic! Perhaps in the future, SMTs can be configured to have a Key and a sum function that have an intrinsic connection.

> Taking a step back, if only the merkle library needs the ability to construct these `ChildError` responses why is it a blocker with the way things are currently where we just use `hex::encode` while constructing the error?

This is definitely possible, and I have a draft PR up for that https://github.com/FuelLabs/fuel-vm/pull/356. The one concern that comes to mind is that `ChildError` can be returned for tree traversal in any sort of tree (BMT, SMT, sum BMT, etc.) where it might not make sense to hex encode other Key types.

**Comment by Voxelot** [comment]
> The one concern that comes to mind is that ChildError can be returned for tree traversal in any sort of tree (BMT, SMT, sum BMT, etc.) where it might not make sense to hex encode the Key type.

One way to address this is to make a wrapper type for keys, such as an enum. You don't need to modify any of the internals of the tree impls to use this type, and could just use it on the ChildError to differentiate between bmt vs smt keys.

**Pull request closed without merging**